### DISCUSSION TITLE
Batching multimeshes

### DISCUSSION THREAD

**Discussion by oeleo1** [godotengine/godot-proposals#9673] [Engine Core]
I have been pondering lately on the idea of multimesh batching and would like to see what others might think of it.

**Background**

The rendering team, @lawnjelly in particular, has done the bulk of the work for batching the most common primitives and there is a sound batching framework in place in Godot's core. But multimeshes are not adressed, they are handled individually and obviously they break batching. Yet, every `CPUParticles/2D` node becomes a multimesh during rendering, as well as `MultiMeshInstance/2D` nodes. A common use case for example is attaching a few growing sparkles here and there, then adding lots of nodes with such sparkles. This produces long sequences of MM batches during frame analysis. A typical output looks like this:
```
	...
	joined_item 1 refs,  ( prevent_vertex_baking, prevent_item_joining,  )
		(z 7)
			batch D 0-1 MM 
	joined_item 1 refs,  ( prevent_vertex_baking, prevent_item_joining,  )
		(z 7)
			batch D 0-1 MM 
	joined_item 1 refs,  ( prevent_vertex_baking, prevent_item_joining,  )
		(z 7)
			batch D 0-1 MM 
	joined_item 1 refs,  ( prevent_vertex_baking, prevent_item_joining,  )
		(z 7)
			batch D 0-1 MM 
	joined_item 1 refs,  ( prevent_vertex_baking, prevent_item_joining,  )
		(z 7)
			batch D 0-1 MM 
	joined_item 1 refs,  ( prevent_vertex_baking, prevent_item_joining,  )
		(z 7)
			batch D 0-1 MM 
	joined_item 1 refs,  ( prevent_vertex_baking, prevent_item_joining,  )
		(z 7)
			batch D 0-1 MM 
	joined_item 1 refs,  ( prevent_vertex_baking, prevent_item_joining,  )
		(z 8)
			batch D 0-1 MM 
	joined_item 1 refs,  ( prevent_vertex_baking, prevent_item_joining,  )
		(z 8)
			batch D 0-1 MM 
	joined_item 1 refs,  ( prevent_vertex_baking, prevent_item_joining,  )
		(z 8)
			batch D 0-1 MM 
	joined_item 1 refs,  ( prevent_vertex_baking, prevent_item_joining,  )
		(z 8)
			batch D 0-1 MM 
	joined_item 1 refs,  ( prevent_vertex_baking, prevent_item_joining,  )
		(z 8)
			batch D 0-1 MM 
	joined_item 1 refs,  ( prevent_vertex_baking, prevent_item_joining,  )
		(z 8)
			batch D 0-1 MM 
	joined_item 1 refs,  ( prevent_vertex_baking, prevent_item_joining,  )
		(z 8)
			batch D 0-1 MM 
	...
```
The nodes have been placed on the same z-index on purpose to allow for better batching of individual node subtrees involving `Sprite` nodes and similar, including `CPUParticles/2D`.

**The idea**

The idea is to reduce the number of draw calls for multimesh instances of the same type. This comes down to batching multimeshes, or to be more precise _to fold_ multimeshes. This means merging legacy multimesh instances in a batched multimesh instance which will be handled by the renderer in only 1 call. In practice this boils down to merging their data buffers in the batcher. The buffer copying may be reduced or not depending on the number of currently visible instances.

I am sure the idea has been played with by others as well, especially for 3D. For example, see [this link](https://www.reddit.com/r/godot/comments/n1oc8a/i_made_an_auto_batching_script_that_reduced_the/) for an interesting app level experiment.

While batching different types of multimeshes is probably doable, it would be much more complicated as a generic mechanism since the involved meshes and materials may be different. Hence, the idea here is to limit multimesh batching only to meshes of the same type: same mesh, same material, same multimesh format (color, transform, custom data). This is much more reasonable, it can be handled by the batcher, and it solves the common scenario quoted above of having the same `CPUParticles/2D` spread all over the place.

I expect that the savings in terms of draw calls will vary depending on the project since folding multimeshes comes at the expense of allocating more memory to merge their buffers. Which is why it would be a good idea, should this be played with, to add it as a batching option with a cap on the size of multimesh instances (for example, batch only multimeshes smaller than *some number*, say 256, instances).

At this point I don't have a good sense of the usefullness of the idea so sharing it will hopefully collect more opinions on its relevance. As to what the impact will be, I guess noone knows unless it is prototyped and tested. Which I may try to do anyway in a great moment of inspiration in order to cut off these long MM draw call sequences.

Thanks.

**Comment by bikemurt** [comment]
This would be excellent for my 3D game as well.

It's something that would seriously make my current game's development better. My game has several LODs for a procedural tree/forest system, but to summarize:

- At the lowest level, each tree is 3D (in fact, each branch is procedurally generated as a part of a multimesh, but that has nothing to do with why this feature would be good for me)
- ![image](https://github.com/user-attachments/assets/a44ebc63-7dea-425a-9e6b-8dd45e2f470f)
- This detailed LOD fades out and transitions into billboards once the camera is far enough
- At this point 1 billboard = 1 tree, and although Godot batching is pretty good at combining these draw calls, there are two things that make it inefficient: 1) the number of objects that the batcher needs to parse through easily gets into the 1000s, 2) I don't need shadows draw for trees/forests way in the background
- ![image](https://github.com/user-attachments/assets/d1441773-57e6-4c89-9d96-52a73cdad69f)
- The lowest level of detail is to use multimeshes of billboards for each chunk of the terrain **(this is where the batching of multi-meshes comes in):**

![image](https://github.com/user-attachments/assets/ceda0939-81d0-46ba-bacb-6594db795dca)

There are 9 tree material variants (same underlying quad mesh), so each chunk of the terrain could have maximum 9 multimeshes, each with 1 draw call (9 total draw calls).

What I'd *like* is for all of these multimesh instances to have batched rendering. Theoretically, it should be possible. The actual multimesh resource on each MultiMeshInstance3D needs to be unique, of course, for the actual positions of the trees to match the position of the terrain, but otherwise they are just the same 9 material instances, all with an underlying Quad mesh.

Edit: I forgot to note that the location of the "X" in the above grid picture is the players grid location: that's where the highest level of detail occurs. So once the player is close enough to a tree, a fully interactable, 3D tree fades in. As the player moves away, the lower levels of LODs kick in.

**Comment by fire** [comment]
Add indirect draw functionality to MultiMesh https://github.com/godotengine/godot/pull/99455 is in 4.4 and can be used for multimesh batching.

Let me know if that is sufficient or is more wanted.

If it's sufficient, we can close this discussion.

**Reply by bikemurt** [reply]
I'll try cherry picking and will update with what I find. Thank you!

**Reply by bikemurt** [reply]
Working through some things, but posting on the PR since the original developer might know the answer: https://github.com/godotengine/godot/pull/99455/#issuecomment-2646553255

**Comment by bikemurt** [comment]
Short update just in case anyone is curious:

- Batching several multimeshes into a single draw call is possible, but not worth it, because it involves re-buffering all of the multimesh transform data on the fly (each time the draw calls are sorted). In my case this is quite a large amount of data since I am dealing with several thousands of instances. Too much memory to be shuffling around at runtime
- The code changes to do so are beyond me (I spent half a day at it, but could not get it to work), but it may not be worth it to pursue anyways, because of the above reasons
- In my case, I ended up doing more of the buffering of the multimesh data in my game level code to be able to combine some multimeshes that were in the background. There's tradeoffs here too: if you combine too many things, then you necessarily have a larger custom AABB, so you may lose some frustum culling

You can have either modularity with your multimeshes at runtime, or fewer draw calls by merging multimeshes, but typically not both.
